src/operator/nn/mkldnn/mkldnn_rnn_impl.h

```diff
@@ -0,0 +1,589 @@
+#include <vector>
+#include "../../rnn_impl.h"
+#include "mkldnn.hpp"
+#include "./mkldnn_base-inl.h"
+
+namespace mxnet {
+namespace op {
+
+namespace mkldnn_rnn_enum {
+  enum RNNModeType {kRnnRelu, kRnnTanh, kLstm, kGru};
+}
+
+template <typename DType>
+mkldnn::memory::data_type GetMKLDNNDataType() {
+  mkldnn::memory::data_type datatype;
+  switch (typeid(DType).name()[0]) {
+    case 'f':
+    case 'd':
+      datatype = mkldnn::memory::data_type::f32;
+      break;
+    case 'i':
+      datatype = mkldnn::memory::data_type::s32;
+      break;
+    case 's':
+      datatype = mkldnn::memory::data_type::s16;
+      break;
+    case 'a':
+      datatype = mkldnn::memory::data_type::s8;
+      break;
+    case 'h':
+      datatype = mkldnn::memory::data_type::u8;
+      break;
+    default:
+      datatype = mkldnn::memory::data_type::data_undef;
+      break;
+  }
+  return datatype;
+}
+
+mkldnn::algorithm GetMKLDNNAlgo(int mode,
+                                int* ngates,
+                                int* nstates) {
+  mkldnn::algorithm algo;
+  switch (mode) {
+    case mkldnn_rnn_enum::kLstm:
+      *ngates = 4;
+      *nstates = 2;
+      algo = algorithm::vanilla_lstm;
+      break;
+    default:
+      LOG(FATAL) << "unsupported RNN mode" << mode;
+      break;
+  }
+  return algo;
+}
+
+void ReorderForWeight(mkldnn::memory src,
+                      mkldnn::memory dst) {
+  auto r = mkldnn::reorder(src, dst);
+  stream(stream::kind::eager).submit({r}).wait();
+}
+
+template <typename DType>
+mkldnn::memory ConcatData(mkldnn::memory::format src_format,
+                          mkldnn::memory::format dst_format,
+                          std::vector<mkldnn::memory::dims> srcs_cds,
+                          mkldnn::memory::dims dst_cds,
+                          mkldnn::memory::data_type mkldnn_dtype,
+                          int concat_dimension,
+                          std::vector<DType*> srcs_data) {
+  auto cpu_engine = CpuEngine::Get()->get_engine();
+  std::vector<mkldnn::memory::primitive_desc> srcs_pd;
+  std::vector<mkldnn::memory> srcs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    auto desc = memory::desc(srcs_cds[i], mkldnn_dtype, src_format);
+    auto mpd = memory::primitive_desc(desc, cpu_engine);
+    auto src_memory = mkldnn::memory({desc, cpu_engine}, srcs_data[i]);
+    srcs_pd.push_back(mpd);
+    srcs.push_back(src_memory);
+  }
+  std::vector<primitive::at> inputs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    inputs.push_back(srcs[i]);
+  }
+
+  auto dst_desc = mkldnn::memory::desc(dst_cds, mkldnn_dtype, dst_format);
+  auto concat_pd = mkldnn::concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  auto dst = mkldnn::memory(concat_pd.dst_primitive_desc());
+  
+  auto c = mkldnn::concat(concat_pd, inputs, dst);
+  stream(stream::kind::eager).submit({c}).wait();
+  return dst;
+}
+
+template <typename DType>
+void MKLDNNRNNForward(bool state_outputs,
```

Comment:
This function is too long to review and maintain. Can you split it into several functions and combine them together here. I guess some blocks can also be shared when we implement backward.